Block eligibility oracle

noamnelke opened the pull request:
I had to skip `TestMultipleNodes` since it fails for now, need to fix and unskip before merging.

Major TODOs:
- [x] Create validator
- [x] Write tests for oracle/validator
- [ ] Properly initialize `ActivationDB`
- [ ] Include eligibility proof in block (?)

VRF Signer related (are we doing any of this right now?):
- [ ] Implement VRF signer/validator
- [ ] Generate and store VRF key-pair

noamnelke committed:
initial block eligibility oracle

almogdepaz commented:
TestMultipleNodes is flaky and I'm going to fix it as part of a pr anyway so you can just leave it skipped

antonlerner commented on oracle/oracle.go:
```diff
@@ -34,8 +36,13 @@ func (bo *localBlockOracle) Register(isHonest bool, pubkey string) {
-func (bo *localBlockOracle) BlockEligible(id mesh.LayerID, pubKey string) bool {
-	return bo.oc.Eligible(uint32(id), bo.committeeSize, pubKey, nil)
+func (bo *localBlockOracle) BlockEligible(layerID mesh.LayerID) ([]mesh.BlockEligibilityProof, error) {
+	eligible := bo.oc.Eligible(uint32(layerID), bo.committeeSize, bo.nodeID.Key, nil)
```
why duplicate code?

antonlerner commented on oracle/blockoracle.go:
```diff
@@ -0,0 +1,88 @@
+	return bo.eligibilityProofs[layerID], nil
+}
+
+func (bo *MinerBlockOracle) calcEligibilityProofs(epochNumber uint64) error {
```
great, add some UT

noamnelke replied:
added

antonlerner commented on oracle/blockoracle.go:
```diff
@@ -0,0 +1,88 @@
+	atx, err := bo.activationDb.GetAtx(atxId)
+	activeSetSize := atx.ActiveSetSize
+
+	numberOfEligibleBlocks := uint32(bo.committeeSize) * uint32(bo.layersPerEpoch) / activeSetSize
```
I need to check if committee size is constant or varies somehow, I'll get back to you

noamnelke replied:
Go ahead and check, it can't hurt.

I've discussed this yesterday with @barakshani and @zalmen and we've agreed that while the *actual* committee size (or avg. size) will change, the value we're aiming for *won't* (and this is what `committeeSize` stands for here).

antonlerner replied:
@zalmen shouldn't it be num of active Ids in last epoch?

noamnelke replied:
I think you're confusing `committeeSize` with `activeSetSize`. The latter is calculated from the ATX's view.

`committeeSize` is the number of blocks we want to have per layer and it's configured to be 200 at the moment. Since we promise each miner to be eligible for a block at least once per epoch, the actual avg. committee size will become bigger if we have more than 806,400 (the number of blocks per epoch) active mining identities.

antonlerner commented on oracle/blockoracle.go:
```diff
@@ -0,0 +1,88 @@
+		log.Error("getting previous atx id failed: %v", err)
+		return err
+	}
+	atxId := atxIDs[len(atxIDs)-1]
```
check that len(atxIDs) > 0

noamnelke replied:
👍🏻

antonlerner commented on crypto/vrf.go:
```diff
@@ -0,0 +1,14 @@
+}
+
+func (s *VRFSigner) Sign(message []byte) []byte {
+	// TODO: implement
```
IMO can be any signature scheme right now

antonlerner commented on cmd/node/node.go:
```diff
@@ -370,12 +374,19 @@ func (app *SpacemeshApp) Start(cmd *cobra.Command, args []string) {
+	//bo := oracle.NewBlockOracleFromClient(oracleClient, int(app.Config.CONSENSUS.NodesPerLayer))
+	nodesPerLayer := app.Config.CONSENSUS.NodesPerLayer
+	layersPerEpoch := app.Config.CONSENSUS.LayersPerEpoch
+	activationDb := &activation.ActivationDb{}      // TODO: initialize properly
```
There is an implementation of this in our code, however it is disconnected from all flows. I think we should first run tests on each component separately and then combine them together

noamnelke committed:
block eligibility validator

noamnelke committed:
first block oracle test

noamnelke committed:
test and handle empty active set

noamnelke committed:
actually validate the vrf 🤭

Thank god for unit tests.

noamnelke committed:
another adversarial test case

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into layer_committee_vrf

noamnelke committed:
receive atx ID from block, validate ATX, fix bug

The bug was that the layer ID was calculated relative to the start of the epoch, so nobody was eligible outside of epoch 0. We now use the absolute layer ID.

almogdepaz commented on sync/syncer.go:
```diff
@@ -17,7 +17,7 @@ import (
 )
 
```
this was intended for internal package use i think we should keep it that way
and if possible from sync point of view the api should be 
func BlockValid(b block) bool
we can discuss this in person

noamnelke replied:
@almogdepaz just so you don't think I ignored this comment - Anton and I discussed it and it will be addressed in a separate PR on Sunday.

The interface you want (and I agree it's the right one) requires putting the eligibility proofs in the block and implementing a bit more logic, so we wanted to separate it from this PR.

antonlerner commented on cmd/node/app_test.go:
```diff
@@ -56,9 +55,10 @@ func (app *AppTestSuite) initMultipleInstances(t *testing.T, numOfInstances int,
 		bo.Register(true, pub.String())
 
-		err := app.apps[i].initServices(pub.String(), n, store, sgn, bo, bo, numOfInstances)
+		err := app.apps[i].initServices(pub.String(), n, store, sgn, bo, nil, bo, numOfInstances) // TODO: pass blockValidator and hareOracle
```
if this fails the test, revert to old blockoracle and test

antonlerner commented on crypto/vrf.go:
```diff
@@ -0,0 +1,26 @@
+)
+
+type VRFSigner struct {
+	privateKey []byte
```
should it be fixed size maybe?

noamnelke replied:
this is a sort of stub, when we implement the signer we can change this

antonlerner commented on mesh/activation.go:
```diff
@@ -76,6 +76,10 @@ func (t ActivationTx) Id() AtxId {
 	return AtxId{crypto.Keccak256Hash(tx)}
 }
 
+func (t ActivationTx) Validate() error {
```
maybe validate should return true false - if valid or not

noamnelke replied:
I like this pattern for validation where no error means it passed and if it failed the error tells you why.  An additional boolean would just reflect that (if there's an error it should be `false`, otherwise `true`).

How is the calling code supposed to handle the result? If the `error` is not `nil` -- also check the `bool`, or just ignore it and assume it's `true`? If the `bool` is `false` and the `error` is `nil` -- I don't get a reason that I can log?

antonlerner commented on oracle/blockeligibilityvalidator.go:
```diff
@@ -0,0 +1,70 @@
+		log.Error("ATX is invalid: %v", err)
+		return false, err
+	}
+	if atxEpochNumber := uint64(atx.LayerIndex) / uint64(v.layersPerEpoch); epochNumber != atxEpochNumber {
```
extract this to a function, converting from layerId and epoch number is something we do quite often

noamnelke committed:
handle no activations for node

antonlerner commented on mesh/mesh.go:
```diff
@@ -344,14 +344,14 @@ func (m *Mesh) AccumulateRewards(rewardLayer LayerID, params RewardConfig) {
 	//todo: check if block producer was eligible?
 	for _, bl := range l.blocks {
-		if _, found := ids[bl.MinerID]; found {
+		if _, found := ids[bl.MinerID.Key]; found {
```
It's not relevant for your PR, but we need to add a todo here that will count number of blocks by miner in case a miner is eligible for more than one in a layer

antonlerner commented on oracle/blockeligibilityvalidator.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	epochBeacon := v.beaconProvider.GetBeacon(epochNumber)
+	message := make([]byte, len(epochBeacon)+binary.Size(epochNumber)+binary.Size(counter))
```
extract serialization to a function

antonlerner commented on crypto/vrf.go:
```diff
@@ -0,0 +1,26 @@
+}
+
+func ValidateVRF(message, signature, publicKey []byte) error {
+	if !bytes.Equal(message, signature) {
```
can use some sig scheme and validate it

noamnelke replied:
done in #789

noamnelke committed:
extract GetEpoch

antonlerner commented on oracle/blockeligibilityvalidator.go:
```diff
@@ -0,0 +1,70 @@
+	}
+	hash := sha256.Sum256(sig)
+	epochOffset := epochNumber * uint64(v.layersPerEpoch)
+	eligibleLayer := mesh.LayerID(binary.LittleEndian.Uint64(hash[:])%uint64(v.layersPerEpoch) + epochOffset)
```
comment and explain or extract calculation to a function

noamnelke replied:
done in #789

noamnelke committed:
add todo in mesh/mesh.go

antonlerner commented on oracle/blockoracle.go:
```diff
@@ -0,0 +1,121 @@
+
+func (bo *MinerBlockOracle) BlockEligible(layerID mesh.LayerID) ([]mesh.BlockEligibilityProof, error) {
+
+	epochNumber := uint64(layerID) / uint64(bo.layersPerEpoch)
```
extract to method

antonlerner commented on oracle/blockoracle.go:
```diff
@@ -0,0 +1,121 @@
+	for counter := uint32(0); counter < numberOfEligibleBlocks; counter++ {
+		message := make([]byte, len(epochBeacon)+binary.Size(epochNumber)+binary.Size(counter))
+		copy(message, epochBeacon)
+		binary.LittleEndian.PutUint64(message[len(epochBeacon):], epochNumber)
```
code duplication, extract to method

noamnelke committed:
extract serializeVRFMessage

antonlerner commented on oracle/blockoracle.go:
```diff
@@ -0,0 +1,121 @@
+	}
+	numberOfEligibleBlocks := uint32(committeeSize) * uint32(layersPerEpoch) / activeSetSize
+	if numberOfEligibleBlocks == 0 {
+		numberOfEligibleBlocks = 1
```
so we've decided to use this patch eventually?

noamnelke committed:
go fmt

noamnelke committed:
Merge remote-tracking branch 'origin/develop' into layer_committee_vrf
